contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,300 @@
+
+![VM](../images/blog/how-we-built-an-app-server/vm.png)
+
+Structuring apps this way provides flexibility to users while maintaining scalability and security. Added features like scheduling and exporting added new functionality users wanted. Caching and improving the config object (enabling statelessness) helped apps scale effectively. Running apps in VMs helped ensure they were secure and isolated. The next step was to put these apps into production.
+
+## Serving apps and managing data at scale
+
+An app is no good if it doesn't integrate into PostHog. Apps needed to integrate with our ingestion pipeline. They needed to handle an inflow of data, scale, manage tasks, and update themselves. There are many parts needed to make this happen.
+
+Below is a basic diagram of our ingestion pipeline. Most of what we care about is the plugin server and the Kafka data flow leading into it.
+
+```mermaid
+graph TD
+    CLIENT[Client Library]
+    DECIDE["/decide API"]
+    CAPTURE[Capture API]
+    PLUGINS[Plugin server]
+    PERSONS["PostgreSQL (persons table)"]
+    Kafka2[Kafka]
+
+    CLIENT -..-> DECIDE
+
+    CLIENT -..-> CAPTURE
+    CAPTURE --> Kafka
+
+    Kafka <-..- PLUGINS
+
+    PLUGINS <--> PERSONS
+    PLUGINS --> Kafka2
+
+    Kafka2 <-..- ClickHouse
+```
+
+### Using Kafka and Redis
+
+Kafka manages our data flows. It helps us batch, split, and manage parallel work for our apps. For example, Kafka helps us batch and retry events when apps call `processEvent`. We use a Kafka topic (which holds messages and events in a logical order) to continue the flow of data.
+
+Another service we use is Redis to cache data and back the queue for apps. Many parts of the app are reusable and don’t need to be rebuilt each time we run the app. There is data we only need to keep around temporarily. Using Redis to cache data like this dramatically improved app performance and lowered the number of app servers we needed to run.
+
+We also use a couple of Redis patterns. First, we use [Redlock](https://redis.io/docs/manual/patterns/distributed-locks/) for the scheduler and job queue consumer. Redlock ensures these processes are only run on one machine because we only want them to run once. If they ran more than once, we’d have duplication and issues. For example, we only need one scheduler to do the basic work of saying what to do at what time. Second, we use [Pub/Sub](https://redis.io/docs/manual/pubsub/) to check for updates to the apps.
+
+We chose these because we used them elsewhere and they have a proven track record of working well at scale. They help the app server integrate with our existing infrastructure.
+
+### Integrating into the plugin server
+
+To ensure apps connected to the rest of the ingestion pipeline, we decided to integrate apps into our plugin server. The plugin server runs processes for ingesting, formatting, and writing data from users to storage. It runs on Node (because of the decision we made in the hackathon about speed and VMs).
```

Comment:
> we decided to integrate apps into our plugin server.

This is totally not true. Apps were called plugins, and the only thing the plugin server did was to run plugins on events. There was nothing to integrate apps into, as the server was built just to run them.

Later on we added "ingestion" (saving to clickhouse/postgres) to save the extra roundtrip back to Python.

> It runs on Node (because of the decision we made in the hackathon about speed and VMs).

We made the decision that we wanted to run custom JS. That required us to run some JS engine, and Node was the only real choice. We would have run `eval` if there was no other choices. Decisions about VMs and other details on how to implement this came after.

Reply:
Rewrote to get the evolution right.